### PR TITLE
Plumbing IV drip

### PR DESCRIPTION
**Pull request by Time-Green**
"Why aren't you working to integrate fermichem?" I wanted to salvage some stuff from the people milker first and turn it into a normal machine.

Introducing: The auto IV drip!

It's an IV drip with plumbing. Receiving chems are instantly put into the bloodstream and are technically INJECTED. You can also set it to 'donate', which drains blood, puts it in an internal buffer and then sends it away into the plumbing net.

I'm not really sure what you can do with donating blood, since you can already dupe blood with mutagen. But hey, the whole idea behind plumbing was to make it integrated into an actual plumbing system, but since that got nuked the entire design philosophy has been "give kids bombs and see what they do with it". 

I'm sure you can put an illuminescent onto one of these, give him a ton of blood regenerating chems, while also draining his blood (liquid electricity) to feed it to another illuminescent in an attempt to create the ultimate lightbulb or something.

I also:
-  Rearranged the plumbing RCD to showcase the common machinery first
- Made changing drip intensity right click (apparently you could change drip speed)
- Fucked IV code because it's kind of old

:cl: Time-Green
add: Adds the auto IV! You ever wanted to feed the clown actual oceans of morphine? You can! You can technically hook up 9 of these to 1 person, so go into your plumbing RCD and have fun! You can also draw blood with it I guess
qol: Changing the drip speed on an IV is now on right click instead of alt
/:cl:

# Why
More plumbing integration good, especially in the recent growth of chemistry as a whole

### CODE REVIEW

**Comment by Pugie** [comment]
PEOPLE MILKER PEOPLE MILKER

**Comment by ATH1909** [comment]
instead of giving criminals brig time as a sec officer, I can now drain out an amount of blood from them proportionate to the severity of their crimes and store it in a giant blood vat of JUSTICE.

10/10 pr

**Comment by ATH1909** [comment]
I mean, I could technically do that before with normal IV drips, but these ones can move the blood directly into the tank!

This'll also let me set up a public hercuri overdosing station for plasmamen, which'll be nice. Actually, plumbing IV poles could serve as an interesting, immobile, "choose your own dosage" alternative to pills/bottles/patches, which is something that could be worth looking into...

**Comment by Time-Green** [comment]
> I mean, I could technically do that before with normal IV drips, but these ones can move the blood directly into the tank!
> 
> This'll also let me set up a public hercuri overdosing station for plasmamen, which'll be nice. Actually, plumbing IV poles could serve as an interesting, immobile, "choose your own dosage" alternative to pills/bottles/patches, which is something that could be worth looking into...

Beside all the jokes, that’d probably be the most common use. Man this could be great with graffity to let people pick the right treatment

**Comment by ilikesaying** [comment]
making criminals pay with their blood has now become even faster

**Review by Twaticus — APPROVED** [review]

**Comment by Tlaltecuhtli** [comment]
can you milk cow with this?

**Comment by Time-Green** [comment]
> can you milk cow with this?

Why would you milk a cow with an IV you sick fuck

**Comment by Pugie** [comment]
Hmm Time-Green has a short memory maybe supporting fermichem burst some brain cells in their brain...

**Comment by Time-Green** [comment]
> Hmm Time-Green has a short memory maybe supporting fermichem burst some brain cells in their brain...

Thalpy/Bramble/Fermi basically did everything, I only changed like two lines to accomodate plumbing. Also people milker didn't work through a needle, this does.

File: code/datums/components/plumbing/IV_drip.dm [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	demand_connects = SOUTH
+	supply_connects = NORTH
+
+	methods = INJECT
+
+/datum/component/plumbing/IV_drip/Initialize()
+	. = ..()
+
+	RegisterSignal(parent, list(COMSIG_IV_ATTACH), .proc/update_attached)
+	RegisterSignal(parent, list(COMSIG_IV_DETACH), .proc/clear_attached)
```
Comment (Mothblocks) [inline code/datums/components/plumbing/IV_drip.dm:12]: Move these to a `RegisterWithParent`, then unregister then in `UnregisterWithParent`.

File: code/datums/components/plumbing/IV_drip.dm [hunk 2]
```diff
@@ -0,0 +1,21 @@
+///Component that feeds whatever it's supplied with to whoever is buckled to it. Can be safely added to whatever like the others, like chairs or something
+/datum/component/plumbing/IV_drip
```
Comment (Mothblocks) [inline code/datums/components/plumbing/IV_drip.dm:2]:
```suggestion
/datum/component/plumbing/iv_drip
```

File: code/datums/components/plumbing/_plumbing.dm [hunk 3]
```diff
@@ -17,6 +17,10 @@
 	var/turn_connects = TRUE
 	///The layer on which we connect. Don't add multiple. If you want multiple layer connects for some reason you can just add multiple components with different layers
 	var/ducting_layer = DUCT_LAYER_DEFAULT
+	//In-case we don't want the main machine to get the reagents, but perhaps whoever is buckled to it
```
Comment (Mothblocks) [inline code/datums/components/plumbing/_plumbing.dm:20]: Add a doc comment explaining what this variable is.

File: code/game/machinery/iv_drip.dm [hunk 4]
```diff
@@ -166,23 +175,46 @@
...
+	SEND_SIGNAL(src, COMSIG_IV_DETACH, attached)
+
+	attached = null
+	update_icon()
+
+/obj/machinery/iv_drip/proc/get_reagent_holder()
+	if(use_internal_storage)
+		return reagents
+	else if(beaker)
+		return beaker.reagents
```
Comment (Mothblocks) [inline code/game/machinery/iv_drip.dm:217]:
```suggestion
	return use_internal_storage ? reagents : beaker?.reagents
```

File: code/game/machinery/iv_drip.dm [hunk 5]
```diff
@@ -166,23 +175,46 @@
...
 	if(dripfeed)
 		dripfeed = FALSE
 		to_chat(usr, "<span class='notice'>You loosen the valve to speed up the [src].</span>")
 	else
 		dripfeed = TRUE
 		to_chat(usr, "<span class='notice'>You tighten the valve to slowly drip-feed the contents of [src].</span>")
+	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+///called when an IV is attached
+/obj/machinery/iv_drip/proc/attach_IV(mob/living/target, mob/usr)
```
Comment (Mothblocks) [inline code/game/machinery/iv_drip.dm:195]:
```suggestion
/obj/machinery/iv_drip/proc/attach_iv(mob/living/target, mob/user)
```

Do not shorten variable names: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/machinery/iv_drip.dm [hunk 6]
```diff
@@ -166,23 +175,46 @@
...
+	log_combat(usr, target, "attached", src, "containing: ([container.log_list()])")
+	add_fingerprint(usr)
+	attached = target
+	START_PROCESSING(SSmachines, src)
+	update_icon()
+
+	SEND_SIGNAL(src, COMSIG_IV_ATTACH, target)
+
+///Called when an iv is detached. doesnt include chat stuff because there's multiple options and its better handled by the caller
+/obj/machinery/iv_drip/proc/detach_IV()
```
Comment (Mothblocks) [inline code/game/machinery/iv_drip.dm:207]:
```suggestion
/obj/machinery/iv_drip/proc/detach_iv()
```

File: code/datums/components/plumbing/IV_drip.dm [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+/datum/component/plumbing/IV_drip/Initialize()
+	. = ..()
+
+	RegisterSignal(parent, list(COMSIG_IV_ATTACH), .proc/update_attached)
+	RegisterSignal(parent, list(COMSIG_IV_DETACH), .proc/clear_attached)
+
+	recipient_reagents_holder = null
+
+/datum/component/plumbing/IV_drip/proc/update_attached(datum/source, mob/living/attachee)
```
Comment (Mothblocks) [inline code/datums/components/plumbing/IV_drip.dm:16]: This, and every other signal handler you make, needs to have `SIGNAL_HANDLER` attached to it.

**Review by Mothblocks — APPROVED** [review]

**Review by Mothblocks — CHANGES_REQUESTED** [review]
err i meant

**Comment by Pugie** [comment]
I still prefer people milker

**Comment by Time-Green** [comment]
I'll probably make a secondary PR removing blood duplication from mutagen and adding an industrial blood replenisher, so you can still farm big quantities of blood for the purpose of synthflesh and such

File: code/datums/components/plumbing/IV_drip.dm [hunk 8]
```diff
@@ -1,21 +1,31 @@
...
 	methods = INJECT
 
-/datum/component/plumbing/IV_drip/Initialize()
+/datum/component/plumbing/iv_drip/RegisterWithParent()
 	. = ..()
 
 	RegisterSignal(parent, list(COMSIG_IV_ATTACH), .proc/update_attached)
 	RegisterSignal(parent, list(COMSIG_IV_DETACH), .proc/clear_attached)
 
 	recipient_reagents_holder = null
```
Comment (Mothblocks) [inline code/datums/components/plumbing/IV_drip.dm:14]: This should probably still be in Initialize, RegisterWithParent is *just* for registering signals.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/components/plumbing/IV_drip.dm (hunk 8 above)
Comment (Time-Green) [reply code/datums/components/plumbing/IV_drip.dm:14]: You’re right, I have no idea what kind of brainfart allowed me to do this

**Comment by TemporalOroboros** [comment]
@Time-Green I thought duping blood with mutagen wiped the blood type making it always toxic.

**Comment by Time-Green** [comment]
> @Time-Green I thought duping blood with mutagen wiped the blood type, making it always toxic.

Not sure, but it's caused by a virology interaction
https://github.com/tgstation/tgstation/blob/4c4ce8bf6200a9f136e1194a4da4c321691ccd45/code/modules/reagents/chemistry/recipes/others.dm#L213

Funnily enough, plasma, uranium, virus food etc also duplicate blood. Either way, removing it later would increase scarcity on blood, letting me buff syntflesh a bit and giving me the opportunity to create some unique blood based meds

**Comment by Mothblocks** [comment]
Make sure to preserve the actual virology purpose 😅

**Comment by Time-Green** [comment]
> Make sure to preserve the actual virology purpose 😅

I’ll try 😈

**Pull request merged**